### PR TITLE
Overall, feels like a good baseline!

### PR DESCRIPTION
**Pull request by matklad**
We have a three angles of attack left I think:

* benchmarks, benchmarks, benchmarks
* switch naive enum representation to an optimized one, which gives  us
  the assembly we want
  * we might want to add an easy way to look at the asm for specific ops?
* polish API -- feature parity, docs, etc

Sending as a PR just to make it easy to discuss. Please don't merge it, and just incorporate suggestions you feel are valuable.

### CODE REVIEW

File: tests/test.rs [hunk 1]
```diff
@@ -16,6 +18,7 @@ fn random_string(n: usize) -> String {
 }
 
 fn is_same(s: &String, fs: &FastString) -> bool {
+    // Why not just `s == fs` ?
```
Comment (MBkkt) [inline tests/test.rs:21]: We want to check that is_empty and len also work

File: tests/test.rs [hunk 2]
```diff
@@ -95,14 +95,17 @@ fn test_remove_prop(mut string: String, index: usize) -> bool {
...
     for command in commands.iter() {
+        // :thumbsup:
         let result = match command {
             Command::Push { ch } => push_command(&mut string, &mut fast_string, *ch),
             Command::PushStr { add } => {
                 push_str_command(&mut string, &mut fast_string, add.as_str())
             }
             Command::Remove { idx } => remove_command(&mut string, &mut fast_string, *idx),
             Command::Clone => {
+                // Do we want to assert some props about clones?
```
Comment (MBkkt) [inline tests/test.rs:108]: I would like to make sure that after cloning the string it remains in the same state and after that it works correctly

File: tests/test.rs [hunk 3]
```diff
@@ -111,9 +114,12 @@ fn test_all_prop(mut string: String, commands: Vec<Command>) -> bool {
             return false;
         }
     }
-    return true;
+
+    true
 }
 
+// Can we remove random tests? quickcheck-based tessts are probably more
```
Comment (MBkkt) [inline tests/test.rs:121]: Maybe yes, maybe no, I'm confused by the fact that quickcheck tests pass very quickly, which means the size of the input is not large

File: tests/test.rs (hunk 1 above)
Comment (matklad) [reply tests/test.rs:21]: Oh, riiight. For some reason I was thinking that short-circuting might play a trick on us here (ie, we return false due to is_empty check,  while `==` would've return `true`). But, clearly, we only short circuit on false, and and false leads to a test failure anyway. 

To put it another way, if theer's a bug in `len` (but not in is_empty or `==`), the current code uncovers it!

**Pull request closed without merging**